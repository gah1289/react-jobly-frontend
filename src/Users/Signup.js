import React, { useState } from 'react';
import JoblyApi from '../api';
import { Button, Form, FormGroup, Label, Input, Col, FormFeedback, Modal, ModalBody } from 'reactstrap';

function Signup() {
	const INITIAL_STATE = {
		username  : '',
		password  : '',
		firstName : '',
		lastName  : '',
		email     : ''
	};

	const testUser = {
		username  : 'gah1289',
		password  : '000000',
		firstName : 'Gabby',
		lastName  : 'McCarthy',
		email     : 'gah@gmail.com'
	};

	const [
		modal,
		setModal
	] = useState(false);

	const [
		error,
		setError
	] = useState();
	const [
		pwError,
		setPwError
	] = useState(false);

	const [
		usernameError,
		setUsernameError
	] = useState(false);

	const [
		formData,
		setFormData
	] = useState(INITIAL_STATE);

	const handleChange = async (e) => {
		const { name, value } = e.target;
		setPwError(false);
		setError();
		setUsernameError(false);
		setFormData((formData) => ({ ...formData, [name]: value }));
	};
	const handleSubmit = async (e) => {
		let { username, password } = formData;
		e.preventDefault();

		// filter out companies by search form data.

		try {
			let res = await JoblyApi.register(formData);
			setModal(true);
		} catch (e) {
			if (e[0].includes('Duplicate')) {
				setUsernameError(true);
				setError('Username already taken');
			}
			if (e[0].includes('password')) {
				setPwError(true);
				setError('Password must be at least 5 characters');
			}
			if (e[0].includes('email')) {
				setPwError(true);
				setError('Please enter a valid email');
			}
		}

		// setFormData(INITIAL_STATE);
	};

	return (
		<div>
			<Modal isOpen={modal}>
				{' '}
				<ModalBody>
					Successfully created profile for {formData.username}! Log in <a href="/login">Here</a>
				</ModalBody>
			</Modal>
			<h1>Register</h1>
			<Form onSubmit={handleSubmit}>
				<FormFeedback>{error}</FormFeedback>
				<FormGroup row>
					<Label sm={2} for="username">
						Username:{' '}
					</Label>
					<Col sm={5}>
						<Input
							name="username"
							invalid={usernameError}
							required
							id="username"
							type="text"
							value={formData.username || ''}
							onChange={handleChange}
						/>
						<FormFeedback>{error}</FormFeedback>
					</Col>
				</FormGroup>
				<FormGroup row>
					<Label sm={2} for="firstName">
						First Name:{' '}
					</Label>
					<Col sm={5}>
						<Input
							name="firstName"
							required
							id="firstName"
							type="text"
							value={formData.firstName || ''}
							onChange={handleChange}
						/>
						<FormFeedback>First Name Required.</FormFeedback>
					</Col>
				</FormGroup>
				<FormGroup row>
					<Label sm={2} for="lastName">
						Last Name:{' '}
					</Label>
					<Col sm={5}>
						<Input
							name="lastName"
							required
							id="lastName"
							type="text"
							value={formData.lastName || ''}
							onChange={handleChange}
						/>
						<FormFeedback>Last Name Required.</FormFeedback>
					</Col>
				</FormGroup>
				<FormGroup row>
					<Label sm={2} for="email">
						Email:{' '}
					</Label>
					<Col sm={5}>
						<Input
							name="email"
							required
							id="email"
							type="email"
							value={formData.email || ''}
							onChange={handleChange}
						/>
						<FormFeedback>Email Required.</FormFeedback>
					</Col>
				</FormGroup>

				<FormGroup row>
					<Label sm={2} for="password">
						Password:{' '}
					</Label>
					<Col sm={5}>
						<Input
							invalid={pwError}
							name="password"
							id="password"
							type="password"
							value={formData.password || ''}
							onChange={handleChange}
						/>
						<FormFeedback>{error}</FormFeedback>
					</Col>
				</FormGroup>
				<Button>Sign Up</Button>
			</Form>
		</div>
	);
}

export default Signup;
