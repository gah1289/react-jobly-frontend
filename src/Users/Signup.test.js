import React from 'react';
import { render, screen } from '@testing-library/react';
import Signup from './Signup';
import { MemoryRouter } from 'react-router-dom';
import ItemContext from '../ItemContext';

describe('Renders Profile', function() {
	let isLoggedIn = false;

	let isLoading = false;

	let user = null;

	it('matches snapshot', function() {
		const { asFragment } = render(
			<MemoryRouter>
				<ItemContext.Provider value={{ isLoading, isLoggedIn, user }}>
					<Signup />
				</ItemContext.Provider>
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('shows Signup Form', function() {
		render(
			<MemoryRouter>
				<ItemContext.Provider value={{ isLoading, isLoggedIn, user }}>
					<Signup />
				</ItemContext.Provider>
			</MemoryRouter>
		);

		const firstNameInput = screen.queryByText(/First Name:/i);
		expect(firstNameInput).toBeInTheDocument();
		expect(firstNameInput.tagName).toBe('LABEL');
		const signupBtn = screen.queryByText(/Sign up/i);
		expect(signupBtn).toBeInTheDocument();
	});
});
