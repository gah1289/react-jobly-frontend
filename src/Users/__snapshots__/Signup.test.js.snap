// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Profile matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Register
    </h1>
    <form
      class=""
    >
      <div
        class="invalid-feedback"
      />
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-2 col-form-label"
          for="username"
        >
          Username: 
        </label>
        <div
          class="col-sm-5"
        >
          <input
            aria-invalid="false"
            class="form-control"
            id="username"
            name="username"
            required=""
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-2 col-form-label"
          for="firstName"
        >
          First Name: 
        </label>
        <div
          class="col-sm-5"
        >
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            required=""
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            First Name Required.
          </div>
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-2 col-form-label"
          for="lastName"
        >
          Last Name: 
        </label>
        <div
          class="col-sm-5"
        >
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            required=""
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Last Name Required.
          </div>
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-2 col-form-label"
          for="email"
        >
          Email: 
        </label>
        <div
          class="col-sm-5"
        >
          <input
            class="form-control"
            id="email"
            name="email"
            required=""
            type="email"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Email Required.
          </div>
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-2 col-form-label"
          for="password"
        >
          Password: 
        </label>
        <div
          class="col-sm-5"
        >
          <input
            aria-invalid="false"
            class="form-control"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
      <button
        class="btn btn-secondary"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
